src/operators/operators.cpp

```diff
@@ -809,6 +810,258 @@ void SumFactorizedOperators<dim,n_faces>::Hadamard_product(
     }
 }
 
+template <int dim, int n_faces>  
+void SumFactorizedOperators<dim,n_faces>::sum_factorized_Hadamard_sparsity_pattern(
```

Comment:
Based on unpublished work, based off the basis structure in reference space, we can solve a HAdamard product in O(n^{d+1}) flops, similar to sum-factorization. The issue with the previous implementation was that it would allocate a matrix of size n^d x n^d then perform O(n^{d+1}) flops with it. When allocating a matrix of size n^d x n^d, that creates an opertaion of size O(n^{2d}). So the new functions creat a pseudo sparsity structure, where we know what the index positions would be in an n^d x n^d matrix without ever having to allocate n^2d. Instead, it allocates an n^d x n sized matrix and fills it with the non-zero elements. These relate to performance because it allows the solver to scale at O(n^{d+1}) completely, for both flops and allocation on-the-fly.

Reply:
Perfect, thanks for the explanation!